newIDE/app/src/VariablesList/VariableRow.js

```diff
@@ -75,71 +84,180 @@ const VariableRow = ({
+          fullWidth
+          name={key + 'name'}
+          defaultValue={name}
+          errorText={errorText}
+          onBlur={onBlur}
+          disabled={origin === 'parent'}
         />
-      </TreeTableCell>
-    );
-  } else {
+      )}
+    </TreeTableCell>,
+  ];
+  if (isCollection) {
     columns.push(
       <TreeTableCell
         expand
         key="value"
         style={limitEditing ? styles.fadedButton : undefined}
       >
-        <Text noMargin>(Structure)</Text>
+        <Text noMargin>
+          ({type === gd.Variable.Structure ? 'Structure' : 'Array'})
+        </Text>
       </TreeTableCell>
     );
+  } else {
+    if (type !== gd.Variable.Boolean)
+      columns.push(
+        <TreeTableCell key="value" expand>
+          <SemiControlledTextField
+            margin="none"
+            type={type === gd.Variable.String ? 'text' : 'number'}
+            commitOnBlur={commitVariableValueOnBlur}
+            fullWidth
+            name={key + 'value'}
+            value={
+              type === gd.Variable.String
+                ? variable.getString()
+                : variable.getValue()
+            }
+            onChange={newValue => {
+              if (
+                type === gd.Variable.String
+                  ? variable.getString() !== newValue
+                  : variable.getValue() !== newValue
+              ) {
+                onChangeValue(newValue);
+              }
+            }}
+            disabled={origin === 'parent' && depth !== 0}
+            multiline={type === gd.Variable.String}
+          />
+        </TreeTableCell>
+      );
+    else
+      columns.push(
+        <TreeTableCell key="value" expand>
+          <BooleanEditor
+            value={variable.getBool()}
+            onToggle={onChangeValue}
+            disabled={origin === 'parent' && depth !== 0}
+          />
+        </TreeTableCell>
+      );
   }
   columns.push(
     <TreeTableCell key="tools" style={styles.toolColumn}>
-      {origin === 'inherited' && depth === 0 && (
+      {origin === 'inherited' && depth === 0 ? (
         <IconButton
           size="small"
           onClick={onResetToDefaultValue}
-          style={isStructure ? undefined : styles.fadedButton}
+          style={isCollection ? undefined : styles.fadedButton}
           tooltip={t`Reset`}
         >
           <Replay />
         </IconButton>
-      )}
-      {!(origin === 'inherited' && depth === 0) && origin !== 'parent' && (
-        <IconButton
-          size="small"
-          onClick={onAddChild}
-          style={isStructure ? undefined : styles.fadedButton}
-          tooltip={t`Add child variable`}
-        >
-          <AddCircle />
-        </IconButton>
+      ) : (
+        origin !== 'parent' && (
+          <>
+            <ElementWithMenu
+              shouldOpen={e =>
+                isCollection && !e.shiftKey && e.type !== 'contextmenu'
+                  ? onAddChild() && false
+                  : true
```

Comment:
This is a bit cryptic, especially the side effect in a ternary ;)
Could we separate the behavior of this button when it's a collection and when not? Like:

{ isCollection ? <ElementWithMenu ... /> : <Whatever ... /> }